README.md
# Name

[Dancer2::Plugin::Pg](https://metacpan.org/pod/Dancer2::Plugin::Pg) - PostgreSQL connection for Dancer2

# SYNOPSIS

    use Dancer2;
    use Dancer2::Plugin::Pg;
    
    my $sth = Pg->query('INSERT INTO table (bar, baz) VALUES (?, ?) RETURNING foo', 'value 1', 'value 2');
    print $sth->fetch->[0];

# CONFIGURATION

    plugins:
        Pg:
            host: 'localhost'
            port: '5432'
            base: 'database'
            username: 'postges'
            password: ''
            options:
                AutoCommit: 1
                AutoInactiveDestroy: 1
                PrintError: 0
                RaiseError: 1
                

Or connections:

    plugins:
        Pg:
            connections:
                default:
                    host: 'localhost'
                    port: '5432'
                    base: 'database1'
                    username: 'postges'
                    password: ''
                foo:
                    host: 'localhost'
                    port: '5432'
                    base: 'database2'
                    username: 'postges'
                    password: ''

default:

    my $pg = Pg;
    
foo:

    my $pg = Pg('foo');
    
# METHODS

### dbh

Return reference connection of DBI

    my $dbh = Pg->dbh;
    $dbh->do(
       q{
           CREATE TABLE table (
              id SERIAL,
              name VARCHAR(50) NOT NULL,
              PRIMARY KEY(id)
           );
       }
    );
    
### query

Method query is main, can be executed insert, update, delete and select.

    my $sth = Pg->query('SELECT * FROM table WHERE id = ?', 7);
    while (my @row = $sth->fetchrow_array) {
       print "@row\n";
    }
    
### selectOne

Method return only column.

    my $total = Pg->selectOne('SELECT COUNT(*) FROM table');
    print $total;
    
### selectRow

Method return row of data fetched.

    my $row = Pg->selectRow('SELECT bar, foo FROM table WHERE id = ?', 7);
    print $row->{bar}, $row->{foo};
    
### selectAll

Method return all rows of data fetched.

    my $all = Pg->selectAll('SELECT bar, foo FROM table');
    while(my $row = $all){
        print $row->{bar}, $row->{foo};
    }
    
### lastInsertID

Method return last insert id

    my $id = Pg->lastInsertID('table', 'column');
    
### column

Method used with methods: insert, update and delete.

    my $pg = Pg;
    $pg->table('foo');
    $pg->column('id', 1);
    $pg->column('name', 'bar');
    $pg->insert;
    
### insert

Method generate SQL and use method query to insert into database.

    my $pg = Pg;
    $pg->table('foo');
    $pg->column('name', 'bar');
    $pg->column('age', 7);
    $pg->returning('id, name'); # method RETURNING PostgreSQL
    my $result = $pg->insert;
    print $result->{id}, $result->{name};
    
### update

Method generate SQL and use method query to update into database.

    my $pg = Pg;
    $pg->table('foo');
    $pg->column('name', 'bar');
    $pg->column('age', {'+=' => 1});
    $pg->returning('id, name'); # method RETURNING PostgreSQL
    my $result = $pg->update(id => 1);
    print $result->{id}, $result->{name};
    
### delete

Method generate SQL and use method query to delete into database.

    my $pg = Pg;
    $pg->table('foo');
    $pg->returning('*'); # method RETURNING PostgreSQL
    my $result = $pg->delete(OR => {age => {'>' => 5}, name => {'LIKE' => '%foo%'}}); # WHERE age > 5 OR name LIKE '%foo%'
    print $result->{foo};

# AUTHOR

Lucas Tiago de Moraes, <lucastiagodemoraes@gmail.com>

# LICENSE AND COPYRIGHT
This program is free software; you can redistribute it and/or modify it under the terms of either: the GNU General Public License as published by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


Changes
Revision history for Dancer2-Plugin-Pg

0.02    2016-04-12/06:16
	Solved problem of dependency

0.01    2016-04-12/04:40
        First version, released on an unsuspecting world.



lib/Dancer2/Plugin/Pg/Core.pm
package Dancer2::Plugin::Pg::Core;

use Moo;
use DBI;

has 'dbh' => (
    is => 'ro',
    writer => 'set_dbh'
);

has ['host', 'base', 'port', 'username', 'password'] => (
    is => 'ro'
);

has 'options' => (
    is => 'ro',
    default => sub {
        {AutoCommit => 1, AutoInactiveDestroy => 1, PrintError => 0, RaiseError => 1}
    }
);

has ['table', 'returning'] => (
    is => 'rw'
);

has ['keys', 'values', 'type', 'reference'] => (
    is => 'rw',
    default => sub {
        []
    }
);

sub BUILD {
    my $self = shift;
    
    my $dsn = 'dbi:Pg:';
    $dsn .= 'dbname=' . $self->base if $self->base;
    $dsn .= ';host=' . $self->host if $self->host;
    $dsn .= ';port=' . $self->port if $self->port;
    
    my $dbh = DBI->connect($dsn, $self->username, $self->password, $self->options) || die $DBI::errstr;
    $self->set_dbh($dbh);
    return $self;
}

sub query {
    my $self = shift;
    
    my $sql = shift;
    my $sth = $self->dbh->prepare($sql) || die $self->dbh->errstr;
    $sth->execute(@_);
    return $sth;
}

sub selectOne {
    my $self = shift;
    
    my $result = $self->dbh->selectrow_arrayref(shift, undef, @_) || die $self->dbh->errstr;
    return $result->[0];
}

sub selectRow {
    my $self = shift;
    
    my $result = $self->dbh->selectrow_hashref(shift, undef, @_) || die $self->dbh->errstr;
    return $result;
}

sub selectAll {
    my $self = shift;
    
    my $result = $self->dbh->selectall_arrayref(shift, {Slice=>{}}, @_) || die $self->dbh->errstr;
    return [] unless $result;
    return [$result] unless ref($result) eq 'ARRAY';
    return $result;
}

sub column {
    my ($self, $key, $value) = @_;
    
    push(@{$self->keys}, $key);
    if (ref($value) eq 'HASH') {
        push(@{$self->type}, keys %{$value});
        push(@{$self->values}, values %{$value});
    }else{
        push(@{$self->type}, undef);
        push(@{$self->values}, $value);
    }
    push(@{$self->reference}, '?');
}

sub insert {
    my $self = shift;
    
    my $sql = 'INSERT INTO ' . $self->table . ' (';
    $sql .= join(',',@{$self->keys});
    $sql .= ') VALUES (';
    $sql .= join(',', @{$self->reference});
    $sql .= ')';
    $sql .= ' RETURNING ' . $self->returning if $self->returning;
    my $sth = $self->query($sql, @{$self->values});
    $self->_clean unless $self->returning;
    return 0 unless $sth;
    if ($self->returning && $sth->rows > 0) {
        return $self->_getReturning($sth, $self->returning);
    }else{
        $self->_clean;
    }
    return $sth || 1;
}

sub update {
    my ($self, %wheres) = @_;
    
    my $sql = undef;
    for(keys @{$self->keys}){
        unless($sql){
            $sql = 'UPDATE ' . $self->table . ' SET ';
            if (${$self->type}[$_]) {
                $sql .= ${$self->keys}[$_] . ' ' . ${$self->type}[$_] . ' ?';
            }else{
                $sql .= ${$self->keys}[$_] . ' = ?';
            }
        }else{
            if (${$self->type}[$_]) {
                $sql .= ', ' . ${$self->keys}[$_] . ' ' . ${$self->type}[$_] . ' ?';
            }else{
                $sql .= ', ' . ${$self->keys}[$_] . ' = ?';
            }
        }
    }
    my $where = '';
    foreach(keys %wheres){
        if ($_ =~ /and|or/i) {
            foreach my $key (keys %{$wheres{$_}}){
                $where .= ' ' . uc($_) if $where;
                if (ref($wheres{$_}{$key}) eq 'HASH') {
                    $where .= ' ' . $key;
                    $where .= ' ' . $_ . ' ?' for(keys %{$wheres{$_}{$key}});
                    push(@{$self->values}, values %{$wheres{$_}{$key}});
                }else{
                    $where .= ' ' . $key . ' = ?';
                    push(@{$self->values}, $wheres{$_}{$key});
                }
            }
        }else{
            $where .= ' AND' if $where;
            if (ref($wheres{$_}) eq 'HASH') {
                $where .= ' ' . $_;
                $where .= ' ' . $_ . ' ?' for(keys %{$wheres{$_}});
                push(@{$self->values}, values %{$wheres{$_}});
            }else{
                $where .= ' ' . $_ . ' = ?';
                push(@{$self->values}, $wheres{$_});
            }
        }
    }
    $sql .= ' WHERE' . $where;
    $sql .= ' RETURNING ' . $self->returning if $self->returning;
    my $sth = $self->query($sql, @{$self->values});
    $self->_clean unless $self->returning;
    return 0 unless $sth;
    if ($self->returning && $sth->rows > 0) {
        return $self->_getReturning($sth, $self->returning);
    }else{
        $self->_clean;
    }
    return $sth || 1;
}

sub delete {
    my ($self, %wheres) = @_;
    
    my $sql = 'DELETE FROM ' . $self->table;
    my $where = '';
    foreach(keys %wheres){
        if ($_ =~ /and|or/i) {
            foreach my $key (keys %{$wheres{$_}}){
                $where .= ' ' . uc($_) if $where;
                if (ref($wheres{$_}{$key}) eq 'HASH') {
                    $where .= ' ' . $key;
                    $where .= ' ' . $_ . ' ?' for(keys %{$wheres{$_}{$key}});
                    push(@{$self->values}, values %{$wheres{$_}{$key}});
                }else{
                    $where .= ' ' . $key . ' = ?';
                    push(@{$self->values}, $wheres{$_}{$key});
                }
            }
        }else{
            $where .= ' AND' if $where;
            if (ref($wheres{$_}) eq 'HASH') {
                $where .= ' ' . $_;
                $where .= ' ' . $_ . ' ?' for(keys %{$wheres{$_}});
                push(@{$self->values}, values %{$wheres{$_}});
            }else{
                $where .= ' ' . $_ . ' = ?';
                push(@{$self->values}, $wheres{$_});
            }
        }
    }
    $sql .= ' WHERE' . $where;
    $sql .= ' RETURNING ' . $self->returning if $self->returning;
    my $sth = $self->query($sql, @{$self->values});
    $self->_clean unless $self->returning;
    return 0 unless $sth;
    if ($self->returning && $sth->rows > 0) {
        return $self->_getReturning($sth, $self->returning);
    }else{
        $self->_clean;
    }
    return $sth || 1;
}

sub lastInsertID {
    my $self = shift;
    
    return $self->dbh->last_insert_id(undef, undef, shift||undef, shift||undef) || die $self->dbh->errstr;
}

sub _getReturning {
    my ($self, $sth, $columns) = @_;
    
    my @keys = split(/\,/, $columns);
    my @values = @{$sth->fetch};
    my %hash;
    for my $id(keys @keys) {
        $keys[$id] =~ s/\s//g;
        $hash{$keys[$id]} = $values[$id];
    }
    $self->_clean;
    return \%hash;
}

sub _clean {
    my $self = shift;
    
    $self->table('');
    $self->returning('');
    $self->keys([]);
    $self->values([]);
    $self->type([]);
    $self->reference([]);
}

1;

__END__

=encoding utf8
 
=head1 NAME

Dancer2::Plugin::Pg::Core

=head1 SYNOPSIS

see L<Dancer2::Plugin::Pg>

=head1 AUTHOR
 
Lucas Tiago de Moraes, C<< <lucastiagodemoraes@gmail.com> >>

=head1 LICENSE AND COPYRIGHT
 
This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.
 
See http://dev.perl.org/licenses/ for more information.

lib/Dancer2/Plugin/Pg.pm
package Dancer2::Plugin::Pg;

use Dancer2::Plugin;
use Dancer2::Plugin::Pg::Core;

our $VERSION = '0.02';

register Pg => sub {
    my $dsl = shift;
    my $connection = shift;
    my $conf;
    if($connection) {
        $conf = plugin_setting()->{connections}->{$connection};
    }else{
        if (plugin_setting()->{connections}->{default}) {
            $conf = plugin_setting()->{connections}->{default};
        }else{
            $conf = plugin_setting();
        }
    }
    
    return Dancer2::Plugin::Pg::Core->new($conf);
};

register_plugin;

1;

__END__

=encoding utf8
 
=head1 NAME

Dancer2::Plugin::Pg - PostgreSQL connection for Dancer2

=head1 SYNOPSIS

    use Dancer2;
    use Dancer2::Plugin::Pg;
    
    my $sth = Pg->query('INSERT INTO table (bar, baz) VALUES (?, ?) RETURNING foo', 'value 1', 'value 2');
    print $sth->fetch->[0];
    
=head1 CONFIGURATION

    plugins:
        Pg:
            host: 'localhost'
            port: '5432'
            base: 'database'
            username: 'postges'
            password: ''
            options:
                AutoCommit: 1
                AutoInactiveDestroy: 1
                PrintError: 0
                RaiseError: 1
            
Or connections:

    plugins:
        Pg:
            connections:
                default:
                    host: 'localhost'
                    port: '5432'
                    base: 'database1'
                    username: 'postges'
                    password: ''
                foo:
                    host: 'localhost'
                    port: '5432'
                    base: 'database2'
                    username: 'postges'
                    password: ''

default:
 
    my $pg = Pg;

foo:
 
    my $pg = Pg('foo');



=head1 METHODS

=head3 dbh

Return reference connection of DBI

    my $dbh = Pg->dbh;
    $dbh->do(
       q{
           CREATE TABLE table (
              id SERIAL,
              name VARCHAR(50) NOT NULL,
              PRIMARY KEY(id)
           );
       }
    );

=head3 query

Method query is main, can be executed insert, update, delete and select.

    my $sth = Pg->query('SELECT * FROM table WHERE id = ?', 7);
    while (my @row = $sth->fetchrow_array) {
       print "@row\n";
    }

=head3 selectOne

Method return only column.

    my $total = Pg->selectOne('SELECT COUNT(*) FROM table');
    print $total;

=head3 selectRow

Method return row of data fetched.

    my $row = Pg->selectRow('SELECT bar, foo FROM table WHERE id = ?', 7);
    print $row->{bar}, $row->{foo};
 
=head3 selectAll

Method return all rows of data fetched.

    my $all = Pg->selectAll('SELECT bar, foo FROM table');
    while(my $row = $all){
        print $row->{bar}, $row->{foo};
    }

=head3 lastInsertID

Method return last insert id

    my $id = Pg->lastInsertID('table', 'column');

=head3 column

Method used with methods: insert, update and delete.

    my $pg = Pg;
    $pg->table('foo');
    $pg->column('id', 1);
    $pg->column('name', 'bar');
    $pg->insert;

=head3 insert

Method generate SQL and use method B<query> to insert into database.

    my $pg = Pg;
    $pg->table('foo');
    $pg->column('name', 'bar');
    $pg->column('age', 7);
    $pg->returning('id, name'); # method RETURNING PostgreSQL
    my $result = $pg->insert;
    print $result->{id}, $result->{name};

=head3 update

Method generate SQL and use method B<query> to update into database.

    my $pg = Pg;
    $pg->table('foo');
    $pg->column('name', 'bar');
    $pg->column('age', {'+=' => 1});
    $pg->returning('id, name'); # method RETURNING PostgreSQL
    my $result = $pg->update(id => 1);
    print $result->{id}, $result->{name};

=head3 delete

Method generate SQL and use method B<query> to delete into database.

    my $pg = Pg;
    $pg->table('foo');
    $pg->returning('*'); # method RETURNING PostgreSQL
    my $result = $pg->delete(OR => {age => {'>' => 5}, name => {'LIKE' => '%foo%'}}); # WHERE age > 5 OR name LIKE '%foo%'
    print $result->{foo};

=head1 AUTHOR
 
Lucas Tiago de Moraes, C<< <lucastiagodemoraes@gmail.com> >>

=head1 LICENSE AND COPYRIGHT
 
This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.
 
See http://dev.perl.org/licenses/ for more information.